Peering hosts

jcorbin opened the pull request:
First steps towards peering changes:
- refactors the channel peering structure to:
  - by an explicit map-like
  - have explicit peer objects
- adds skeleton peer state objects, only used for selection in static ways presently
- adds support for hosts option
  - [ ] TODO document
  - [x] TODO tests
  - [ ] TODO example usage

reviewers: @kriskowal @Raynos

jcorbin commented:
ready for review @kriskowal / @Raynos I want to get this in, then start deving sub channel support by integration with autobahn

kriskowal commented:
:+1:

Raynos commented on node/index.js:
```diff
@@ -867,4 +759,259 @@ function TChannelClientOp(req, start) {
+        var score = peer.state.shouldRequest(options, op);
+        var want = score > threshold &&
+                   (selectedPeer === null || score > selectedScore);
+        self.logger.debug('choose peer score', {
```
This happens for every `request()`; consider using `.trace()`

Raynos commented on node/index.js:
```diff
@@ -867,4 +759,259 @@ function TChannelClientOp(req, start) {
+    // TODO: op isn't quite right currently as a "TChannelClientOp", the
+    // intention is that the other (non-options) arg encapsulates all requests
+    // across retries and setries
+    return 0;
```
What does `0` means ? I would guess that `0` means that it should never request.

Raynos commented:
#LEGIT

jcorbin commented:
Updated:
- renamed / implemented more map-like keys/values/entries/delete
- added addPeer, needed in upcoming sub-channel work
- commented out the score debugging for now, I don't think even trace is right

Raynos commented:
lgtm :D
